lightning/src/chain/package.rs

```diff
@@ -0,0 +1,1037 @@
+			},
+			PackageSolvingData::HolderHTLCOutput(ref holder_outp) => {
+				4u8.write(writer)?;
+				holder_outp.write(writer)?;
+			},
+			PackageSolvingData::HolderFundingOutput(ref funding_outp) => {
+				5u8.write(writer)?;
+				funding_outp.write(writer)?;
+			}
+		}
+		Ok(())
+	}
+}
+
+impl Readable for PackageSolvingData {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		let byte = <u8 as Readable>::read(reader)?;
+		let solving_data = match byte {
+			0 => {
+				PackageSolvingData::RevokedOutput(Readable::read(reader)?)
+			},
+			1 => {
+				PackageSolvingData::RevokedHTLCOutput(Readable::read(reader)?)
+			},
+			2 => {
+				PackageSolvingData::CounterpartyOfferedHTLCOutput(Readable::read(reader)?)
+			},
+			3 => {
+				PackageSolvingData::CounterpartyReceivedHTLCOutput(Readable::read(reader)?)
+			},
+			4 => {
+				PackageSolvingData::HolderHTLCOutput(Readable::read(reader)?)
+			},
+			5 => {
+				PackageSolvingData::HolderFundingOutput(Readable::read(reader)?)
+			}
+			_ => return Err(DecodeError::UnknownVersion)
+		};
+		Ok(solving_data)
+	}
+}
+
+/// A malleable package might be aggregated with other packages to save on fees.
+/// A untractable package has been counter-signed and aggregable will break cached counterparty
+/// signatures.
+#[derive(Clone, PartialEq)]
+pub(crate) enum PackageMalleability {
+	Malleable,
+	Untractable,
+}
+
+/// A structure to describe a package content that is generated by ChannelMonitor and
+/// used by OnchainTxHandler to generate and broadcast transactions settling onchain claims.
+///
+/// A package is defined as one or more transactions claiming onchain outputs in reaction
+/// to confirmation of a channel transaction. Those packages might be aggregated to save on
+/// fees, if satisfaction of outputs's witnessScript let's us do so.
+///
+/// As packages are time-sensitive, we fee-bump and rebroadcast them at scheduled intervals.
+/// Failing to confirm a package translate as a loss of funds for the user.
+#[derive(Clone, PartialEq)]
+pub struct PackageTemplate {
+	// List of onchain outputs and solving data to generate satisfying witnesses.
+	inputs: Vec<(BitcoinOutPoint, PackageSolvingData)>,
+	// Packages are deemed as malleable if we have local knwoledge of at least one set of
+	// private keys yielding a satisfying witnesses. Malleability implies that we can aggregate
+	// packages among them to save on fees or rely on RBF to bump their feerates.
+	// Untractable packages have been counter-signed and thus imply that we can't aggregate
+	// them without breaking signatures. Fee-bumping strategy will also rely on CPFP.
+	malleability: PackageMalleability,
+	// Block height after which the earlier-output belonging to this package is mature for a
+	// competing claim by the counterparty. As our chain tip becomes nearer from the timelock,
+	// the fee-bumping frequency will increase. See `OnchainTxHandler::get_height_timer`.
+	absolute_timelock: u32,
```

Comment:
This name really confused me - `absolute_timelock` implies the timelock on a transaction (presumably the transaction being created), but this variable really means "the time at which our counterparty can potentially spend the claims here", maybe call it something like `soonest_conf_deadline` or something?